apm-agent-plugins/apm-es-restclient-plugin/apm-es-restclient-plugin-common/src/main/java/co/elastic/apm/agent/es/restclient/ElasticsearchRestClientInstrumentationHelperImpl.java

```diff
@@ -0,0 +1,145 @@
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.es.restclient;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import co.elastic.apm.agent.objectpool.Allocator;
+import co.elastic.apm.agent.objectpool.ObjectPool;
+import co.elastic.apm.agent.objectpool.impl.QueueBasedObjectPool;
+import co.elastic.apm.agent.util.IOUtils;
+import org.apache.http.HttpEntity;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseException;
+import org.elasticsearch.client.ResponseListener;
+import org.jctools.queues.atomic.AtomicQueueFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+import static org.jctools.queues.spec.ConcurrentQueueSpec.createBoundedMpmc;
+
+public class ElasticsearchRestClientInstrumentationHelperImpl implements ElasticsearchRestClientInstrumentationHelper<HttpEntity, Response, ResponseListener> {
+
+    public static final String SEARCH_QUERY_PATH_SUFFIX = "_search";
+    public static final String SPAN_TYPE = "db";
+    public static final String ELASTICSEARCH = "elasticsearch";
+    public static final String SPAN_ACTION = "request";
+    private static final int MAX_POOLED_ELEMENTS = 256;
+    private final ElasticApmTracer tracer;
+
+    private final ObjectPool<ResponseListenerWrapper> responseListenerObjectPool;
+
+    public ElasticsearchRestClientInstrumentationHelperImpl(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        responseListenerObjectPool = QueueBasedObjectPool.ofRecyclable(
+            AtomicQueueFactory.<ResponseListenerWrapper>newQueue(createBoundedMpmc(MAX_POOLED_ELEMENTS)),
+            false,
+            new ResponseListenerAllocator());
+    }
+
+    private class ResponseListenerAllocator implements Allocator<ResponseListenerWrapper> {
+        @Override
+        public ResponseListenerWrapper createInstance() {
+            return new ResponseListenerWrapper(ElasticsearchRestClientInstrumentationHelperImpl.this);
+        }
+    }
+
+    @Override
+    @Nullable
+    public Span createClientSpan(String method, String endpoint, @Nullable HttpEntity httpEntity) {
+        final TraceContextHolder<?> activeSpan = tracer.getActive();
+        if (activeSpan == null || !activeSpan.isSampled()) {
+            return null;
+        }
+
+        // Don't record nested spans. In 5.x clients the instrumented sync method is calling the instrumented async method
+        if (activeSpan instanceof Span && ELASTICSEARCH.equals(((Span) activeSpan).getSubtype())) {
```

Comment:
Maybe not for now but I think we should add a `external` flag to spans. The tracer should return `null` when trying to create a child span of an external span.

Reply:
I agree, but in this case this is to avoid nested ES client spans, so I would leave it anyway as it deals with a specific problem.